File: 2025-04-April-GFG-POTD/April 23 - Unique Number II/Solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<int> singleNum(vector<int>& arr) {
        int xor_all = 0;

        // Step 1: XOR all elements
        for (int num : arr) {
            xor_all ^= num;
        }

        // Step 2: Find rightmost set bit
        int setBit = xor_all & -xor_all;

        int num1 = 0, num2 = 0;

        // Step 3: Divide into two groups and XOR separately
        for (int num : arr) {
            if (num & setBit) {
                num1 ^= num;
            } else {
                num2 ^= num;
            }
        }

        // Step 4: Return in increasing order
        vector<int> result = {num1, num2};
        sort(result.begin(), result.end());
        return result;
    }
};

int main() {
    int t;
    cin >> t;
    cin.ignore(); // To skip the newline after reading t

    while (t--) {
        string line;
        getline(cin, line);

        vector<int> arr;
        int num;
        stringstream ss(line);
        while (ss >> num) {
            arr.push_back(num);
        }

        Solution sol;
        vector<int> result = sol.singleNum(arr);
        for (int val : result) {
            cout << val << " ";
        }
        cout << endl << "~" << endl;
    }

    return 0;
}

File: 2025-04-April-GFG-POTD/April 23 - Unique Number II/Solution.py
class Solution:
    def singleNum(self, arr):
        xor_all = 0

        # Step 1: XOR all elements
        for num in arr:
            xor_all ^= num

        # Step 2: Find rightmost set bit
        set_bit = xor_all & -xor_all

        num1 = 0
        num2 = 0

        # Step 3: Divide into two groups and XOR separately
        for num in arr:
            if num & set_bit:
                num1 ^= num
            else:
                num2 ^= num

        # Step 4: Return in increasing order
        return sorted([num1, num2])


# Driver code
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().split()))
        sol = Solution()
        result = sol.singleNum(arr)
        print(*result)
        print("~")
